File: augur-engine/src/assets/imagegen.ts
import { writeFile, mkdir } from 'node:fs/promises';
import { join, dirname } from 'node:path';

const REPLICATE_API = 'https://api.replicate.com/v1/models/black-forest-labs/flux-2-klein-4b/predictions';
const FAL_API = 'https://queue.fal.run/fal-ai/flux-2/klein/4b';

interface ReplicateResponse {
  id: string;
  status: string;
  output?: string[];
  error?: string;
  urls?: { get: string };
}

interface FalResponse {
  images?: Array<{ url: string }>;
  request_id?: string;
}

/** Generate an image via Replicate (primary) with fal.ai fallback */
export async function generateImage(
  prompt: string,
  outputPath: string,
): Promise<string> {
  // Try Replicate first
  const replicateToken = process.env['REPLICATE_API_TOKEN'];
  if (replicateToken) {
    try {
      const result = await generateViaReplicate(prompt, replicateToken);
      await downloadAndSave(result, outputPath);
      console.log('[image] generated via Replicate');
      return outputPath;
    } catch (err) {
      console.warn('[image] Replicate failed, trying fal.ai fallback:', err instanceof Error ? err.message : err);
    }
  }

  // Fallback to fal.ai
  const falKey = process.env['FAL_KEY'];
  if (falKey) {
    try {
      const result = await generateViaFal(prompt, falKey);
      await downloadAndSave(result, outputPath);
      console.log('[image] generated via fal.ai (fallback)');
      return outputPath;
    } catch (err) {
      throw new Error(`Both image providers failed. fal.ai error: ${err instanceof Error ? err.message : err}`);
    }
  }

  throw new Error('No image generation API key configured (REPLICATE_API_TOKEN or FAL_KEY)');
}

/** Generate via Replicate API (synchronous prediction) */
async function generateViaReplicate(prompt: string, token: string): Promise<string> {
  // Create prediction
  const createRes = await fetch(REPLICATE_API, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      input: {
        prompt,
        width: 1024,
        height: 768,
        num_outputs: 1,
        output_format: 'webp',
        output_quality: 85,
      },
    }),
  });

  if (!createRes.ok) {
    throw new Error(`Replicate create error: ${createRes.status} ${await createRes.text()}`);
  }

  let prediction = (await createRes.json()) as ReplicateResponse;

  // Poll for completion (max 60s)
  const pollUrl = prediction.urls?.get ?? `https://api.replicate.com/v1/predictions/${prediction.id}`;
  const deadline = Date.now() + 60_000;

  while (prediction.status !== 'succeeded' && prediction.status !== 'failed') {
    if (Date.now() > deadline) throw new Error('Replicate prediction timed out');
    await new Promise((r) => setTimeout(r, 1000));

    const pollRes = await fetch(pollUrl, {
      headers: { 'Authorization': `Bearer ${token}` },
    });
    prediction = (await pollRes.json()) as ReplicateResponse;
  }

  if (prediction.status === 'failed') {
    throw new Error(`Replicate prediction failed: ${prediction.error}`);
  }

  const imageUrl = prediction.output?.[0];
  if (!imageUrl) throw new Error('No image URL in Replicate output');
  return imageUrl;
}

/** Generate via fal.ai API */
async function generateViaFal(prompt: string, key: string): Promise<string> {
  const res = await fetch(FAL_API, {
    method: 'POST',
    headers: {
      'Authorization': `Key ${key}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      prompt,
      image_size: { width: 1024, height: 768 },
      num_images: 1,
      output_format: 'webp',
      sync_mode: true,
    }),
  });

  if (!res.ok) {
    throw new Error(`fal.ai error: ${res.status} ${await res.text()}`);
  }

  const data = (await res.json()) as FalResponse;
  const imageUrl = data.images?.[0]?.url;
  if (!imageUrl) throw new Error('No image URL in fal.ai output');
  return imageUrl;
}

/** Download an image URL and save to disk */
async function downloadAndSave(url: string, outputPath: string): Promise<void> {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Failed to download image: ${res.status}`);

  const buffer = Buffer.from(await res.arrayBuffer());
  await mkdir(dirname(outputPath), { recursive: true });
  await writeFile(outputPath, buffer);
}

File: augur-engine/src/assets/imagegen.py
"""Image generation via Replicate (primary) with fal.ai fallback."""

from __future__ import annotations

import asyncio
import logging
import os
from pathlib import Path

import httpx

log = logging.getLogger("augur.assets")

REPLICATE_API = "https://api.replicate.com/v1/models/black-forest-labs/flux-2-klein-4b/predictions"
FAL_API = "https://queue.fal.run/fal-ai/flux-2/klein/4b"


async def generate_image(prompt: str, output_path: str) -> str:
    """Generate an image via Replicate (primary) with fal.ai fallback.

    Returns the output path on success.
    """
    replicate_token = os.environ.get("REPLICATE_API_TOKEN")
    if replicate_token:
        try:
            url = await _generate_via_replicate(prompt, replicate_token)
            await _download_and_save(url, output_path)
            log.info("generated via Replicate")
            return output_path
        except Exception as exc:
            log.warning("Replicate failed, trying fal.ai fallback: %s", exc)

    fal_key = os.environ.get("FAL_KEY")
    if fal_key:
        try:
            url = await _generate_via_fal(prompt, fal_key)
            await _download_and_save(url, output_path)
            log.info("generated via fal.ai (fallback)")
            return output_path
        except Exception as exc:
            raise RuntimeError(
                f"Both image providers failed. fal.ai error: {exc}"
            ) from exc

    raise RuntimeError(
        "No image generation API key configured (REPLICATE_API_TOKEN or FAL_KEY)"
    )


async def _generate_via_replicate(prompt: str, token: str) -> str:
    """Generate via Replicate API with polling."""
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            REPLICATE_API,
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
            },
            json={
                "input": {
                    "prompt": prompt,
                    "width": 1024,
                    "height": 768,
                    "num_outputs": 1,
                    "output_format": "webp",
                    "output_quality": 85,
                }
            },
        )
        resp.raise_for_status()
        prediction = resp.json()

    poll_url = prediction.get("urls", {}).get(
        "get", f"https://api.replicate.com/v1/predictions/{prediction['id']}"
    )

    async with httpx.AsyncClient(timeout=30) as client:
        for _ in range(60):  # max 60s
            if prediction["status"] in ("succeeded", "failed"):
                break
            await asyncio.sleep(1)
            resp = await client.get(
                poll_url, headers={"Authorization": f"Bearer {token}"}
            )
            prediction = resp.json()

    if prediction["status"] == "failed":
        raise RuntimeError(f"Replicate prediction failed: {prediction.get('error')}")

    output = prediction.get("output", [])
    if not output:
        raise RuntimeError("No image URL in Replicate output")
    return output[0]


async def _generate_via_fal(prompt: str, key: str) -> str:
    """Generate via fal.ai API."""
    async with httpx.AsyncClient(timeout=60) as client:
        resp = await client.post(
            FAL_API,
            headers={
                "Authorization": f"Key {key}",
                "Content-Type": "application/json",
            },
            json={
                "prompt": prompt,
                "image_size": {"width": 1024, "height": 768},
                "num_images": 1,
                "output_format": "webp",
                "sync_mode": True,
            },
        )
        resp.raise_for_status()
        data = resp.json()

    images = data.get("images", [])
    if not images:
        raise RuntimeError("No image URL in fal.ai output")
    return images[0]["url"]


async def _download_and_save(url: str, output_path: str) -> None:
    """Download an image URL and save to disk."""
    Path(output_path).parent.mkdir(parents=True, exist_ok=True)
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.get(url)
        resp.raise_for_status()
        Path(output_path).write_bytes(resp.content)
